File: ciao-prolog/ciao-prolog-1.15.0+r14995/ciao/library/apll/visandor.c
/* Copyright (C) 2007,2008 UPM-CLIP */

#include "datadefs.h"
#include "support.h"

#include <sys/time.h>

#include "visandor.h"


/* ***************************************************************** */
/*                       VISANDOR PRIMITIVES                         */
/* ***************************************************************** */


/* Marks the execution of the local goal as started (VisAndOr event) */
bool_t apll_mark_start_goal_local(Arg)
     Argdecl;
{
#if defined(ANDPARALLEL) && defined(VISANDOR)
  EVENT(START_GOAL, Pcall_Level(w), 0);
#endif

  return TRUE;
}


/* Marks the execution of the goal as started (VisAndOr event) */
bool_t apll_mark_start_goal(Arg)
     Argdecl;
{
#if defined(ANDPARALLEL) && defined(VISANDOR)
  par_handler_t *h = (par_handler_t *) read_handler(X(0));

  if (h != NULL)
    EVENT(START_GOAL, h->ppf, 1);
#endif

  return TRUE;
}


/* Marks the execution of the local goal as finished (VisAndOr
   event) */
bool_t apll_mark_finish_goal_local(Arg)
     Argdecl;
{
#if defined(ANDPARALLEL) && defined(VISANDOR)
  EVENT(FINISH_GOAL, Pcall_Level(w), 0);
#endif

  return TRUE;
}


/* Marks the execution of the goal as finished (VisAndOr event) */
bool_t apll_mark_finish_goal(Arg)
     Argdecl;
{
#if defined(ANDPARALLEL) && defined(VISANDOR)
  par_handler_t *h = (par_handler_t *) read_handler(X(0));

  if (h != NULL)
    EVENT(FINISH_GOAL, h->ppf, 1);
#endif

  return TRUE;
}


/* Marks a join (VisAndOr event) */
bool_t apll_mark_join(Arg)
     Argdecl;
{
#if defined(ANDPARALLEL) && defined(VISANDOR)
  EVENT(JOIN, Pcall_Level(w), 2);
  (Pcall_Level(w))--;
#endif

  return TRUE;
}


/* Starts the event trace */
bool_t apll_start_event_trace(Arg)
     Argdecl;
{
#if defined(ANDPARALLEL) && defined(VISANDOR)
  worker_t *aux = w;

#if !defined(USCLK_EXISTS) && !defined(NSCLK_EXISTS)
  // Initialize the count of calls to usertime(). MCL.
  count_calls = 0;
#endif

  time_at_event_start = (float)0;
  time_at_event_start = (TIME)/1e6;
  do {
    Pcall_Level(aux) = (ENG_INT)aux;
    aux = Next_Wam_Of(aux);
  }
  while (aux!=w);
  gen_event_file = TRUE;
  EVENT(START_TIME,Pcall_Level(w),0);
  EVENT(AGENT_BUSY,0,nacagents);
#if defined(DEBUG)
  if (gen_event_file)
    printf("Events started.\n");
#endif

#endif
  return TRUE;
}


/* Stops the event trace */
bool_t apll_stop_event_trace(Arg)
     Argdecl;
{
#if defined(ANDPARALLEL) && defined(VISANDOR)
  EVENT(AGENT_IDLE,0,nacagents);
  EVENT(STOP_TIME,Pcall_Level(w),0);
#if defined(DEBUG)
  if (gen_event_file)
    printf("Events stopped.\n");
#endif
  gen_event_file = FALSE;
#endif
  return TRUE;
}


/* Saves all events to file in chronological order, clear event tables
   inefficient, but it works */
bool_t apll_save_event_trace(Arg)
     Argdecl;
{
#if defined(ANDPARALLEL) && defined(VISANDOR)
  stream_node_t *stream;
  FILE * event_file;
  visandor_event_t * eventptr;
  int agent;
  worker_t ** current_event;
  int min_ts_agent;
  unsigned long min_tstamp;
  int nagents_finished;
  worker_t *aux = w;

  stream = (stream_node_t *) stream_to_ptr(X(0), 'x');
  if (stream==NULL)
    SERIOUS_FAULT("save_event_trace/1: bad stream argument.");
  event_file = stream->streamfile;
  fprintf(event_file, "0\n");  // 0 means AND parallelism
  current_event =
    (worker_t **) checkalloc(nagents * sizeof(worker_t *));

  // Termination
  nagents_finished = 0;

  // Setup current_event pointers
  // Make timestamp of last event large so that we don't output it
  agent = 0;
  do {
    current_event[agent] = aux;
    (NextEvent(aux))->timestamp = -1;  // Largest unsigned number
    if (FirstEvent(aux) == NextEvent(aux))
      nagents_finished++;
    agent++;
    aux = Next_Wam_Of(aux);
  }
  while (aux!=w);

  do {
    // Find first (min-tstamp) of the current events
    for (min_ts_agent = 0,
           min_tstamp = (FirstEvent((current_event[0])))->timestamp,
	   agent = 1;
	 agent < nagents;
	 agent++) {
      if ( (FirstEvent((current_event[agent])))->timestamp < min_tstamp ) {
	min_ts_agent = agent;
	min_tstamp = (FirstEvent((current_event[agent])))->timestamp;
      }
    }
    // Output it
    eventptr = (FirstEvent((current_event[min_ts_agent])))++;
    fprintf(event_file, "%8lu %d %X %d %X %d \n",
	    eventptr->timestamp, eventptr->evtype, eventptr->ppf,
	    eventptr->nfork, eventptr->wam, min_ts_agent);
    // If we have finished all events for this agent, record it
    if ( (FirstEvent((current_event[min_ts_agent]))) ==
	 (NextEvent((current_event[min_ts_agent]))) ) {
      nagents_finished++;
    }
  }
  while ( nagents_finished < nagents );

  // Reset event tables
  aux = Next_Wam_Of(w);
  do {
    NextEvent(aux) = FirstEvent(aux);
    aux = Next_Wam_Of(aux);
  }
  while (aux!=w);

  return TRUE;
#else
  fprintf(stderr, "[Warning: compiled with AND_PARALLEL_EXECUTION option");
  fprintf(stderr, " not set to \"visandor\", no trace produced]");
  return TRUE;
#endif
}

File: ciao-prolog/ciao-prolog-1.15.0+r14995/ciao/library/apll/visandor.pl
:- module(visandor,
	[
	    mark_start_goal_local/0,
	    mark_start_goal/1,
	    mark_finish_goal_local/0,
	    mark_finish_goal/1,
	    mark_join/0,
	    start_event_trace/0,
	    stop_event_trace/0,
	    save_event_trace/1
        ],
	[assertions, isomodes, foreign_interface]).


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%% VISANDOR PRIMITIVES
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


:- true pred '$mark_start_goal_local'
   + foreign_low(apll_mark_start_goal_local).
:- true pred '$mark_start_goal'(+int)
   + foreign_low(apll_mark_start_goal).
:- true pred '$mark_finish_goal_local'
   + foreign_low(apll_mark_finish_goal_local).
:- true pred '$mark_finish_goal'(+int)
   + foreign_low(apll_mark_finish_goal).
:- true pred '$mark_join'
   + foreign_low(apll_mark_join).
:- true pred '$start_event_trace'
   + foreign_low(apll_start_event_trace).
:- true pred '$stop_event_trace'
   + foreign_low(apll_stop_event_trace).
:- true pred '$save_event_trace'(+int)
   + foreign_low(apll_save_event_trace).


:- pred mark_start_goal_local
     # "Marks the execution of the local goal as started.".

mark_start_goal_local:-
	'$mark_start_goal_local'.

:- pred mark_start_goal(+Handler) :
   int
     # "Marks the execution of the goal as started.".

mark_start_goal(Handler):-
	'$mark_start_goal'(Handler).

:- pred mark_finish_goal_local
     # "Marks the execution of the local goal as finished.".

mark_finish_goal_local:-
	'$mark_finish_goal_local'.

:- pred mark_finish_goal(+Handler) :
   int
     # "Marks the execution of the goal as finished.".

mark_finish_goal(Handler):-
	'$mark_finish_goal'(Handler).

:- pred mark_join
     # "Marks the join of goals.".

mark_join:-
	'$mark_join'.

:- pred start_event_trace
     # "Starts the event trace for VisAndOr.".

start_event_trace:-
	'$start_event_trace'.

:- pred stop_event_trace
     # "Stops the event trace for VisAndOr.".

stop_event_trace:-
	'$stop_event_trace'.

:- pred save_event_trace(+Stream) # "Saves the event trace for VisAndOr with stream @var{Stream} associated to the file.".

save_event_trace(Stream):-
	'$save_event_trace'(Stream).


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


:- use_foreign_source(apll).
:- use_foreign_source(visandor).

:- extra_compiler_opts(['-g']).
:- extra_linker_opts(['-lpthread']).



